assimilation_code/location/threed_cartesian/location_mod.f90

```diff
@@ -1020,30 +1178,72 @@ subroutine get_close(gc, base_loc, base_type, locs, loc_qtys, &
+! specified.  if this is absolutely needed by someone, i can rejigger the
+! code to try to use the default cutoff for identity obs, but it's tricky
+! to identify which obs type is using the default.  in theory it's possible
+! to specify a default cutoff and then specify a per-type cutoff for every
+! other type in the system.  in that case i don't have a map entry for the
+! default, and it's a pain to construct one.  so i'm punting for now.
+if (base_type < 0) then
+   if (gc%nt > 1) then
+      write(msgstring,  '(A)') 'no support for identity obs if per-obs-type cutoffs are specified'
+      write(msgstring1, '(A)') 'contact dart support if you have a need for this combination'
+      call error_handler(E_ERR, 'get_close', msgstring, source, text2=msgstring1)
+   endif
+   bt = 1
+else
+   ! map from type index to gtt index
+   if (base_type < 1 .or. base_type > size(gc%type_to_cutoff_map)) then
+      write(msgstring,'(A,I8)')'base_type out of range, is ', base_type
+      write(msgstring1,'(A,2I8)')'must be between ', 1, size(gc%type_to_cutoff_map)
+      call write_location (0, base_loc, charstring=msgstring2)
+      call error_handler(E_ERR, 'get_close', msgstring, source, &
+                         text2=msgstring1, text3=msgstring2)
+   endif
+   bt = gc%type_to_cutoff_map(base_type)
+   if (bt < 1 .or. bt > gc%nt) then
+      write(msgstring,'(A,I8)')'mapped type index out of range, is ', bt
+      write(msgstring1,'(A,2I8)')'must be between ', 1, gc%nt
+      write(msgstring2, '(A)')'internal error, should not happen.  Contact DART Support'
+      call error_handler(E_ERR, 'get_close', msgstring, source, &
+                         text2=msgstring1, text3=msgstring2)
+   endif
+endif
+
+
+
 ! the list of locations in the loc() argument must be the same
 ! as the list of locations passed into get_close_init(), so
 ! gc%num and size(loc) better be the same.   if the list changes,
 ! you have to destroy the old gc and init a new one.
-if (size(locs) /= gc%num) then
-   write(errstring,*)'locs() array must match one passed to get_close_init()'
-   call error_handler(E_ERR, 'get_close_boxes', errstring, source)
+if (size(locs) /= gc%box(bt)%num) then
+   !write(errstring,*)'locs() array must match one passed to get_close_init()'
+   !call error_handler(E_ERR, 'get_close_boxes', errstring, source)
+   write(msgstring,'(A)')'locs() array must match one passed to get_close_init()'
+   write(msgstring1,'(A,2I8)')'init size, current list size: ', gc%box(bt)%num, size(locs)
+   write(msgstring2,'(A,I8)')'bt = ', bt
+   call error_handler(E_ERR, 'get_close', msgstring, source, &
+                      text2=msgstring1, text3=msgstring2)
 endif
 
 ! If num == 0, no point in going any further.
-if (gc%num == 0) return
+if (gc%box(bt)%num == 0) return
 
-this_maxdist = gc%maxdist
+this_maxdist = gc%box(bt)%maxdist
 
 !> @todo this is doing an exhaustive search each time.  expensive
 !> but should give the right answer.
-
 if(.true.) then
```

Comment:
is this really intended to be .true. ??  this code is going to call exhaustive_collect() every time and not use any of the code in the rest of this routine.  should this be .false. when not debugging?  or true when one or more of the dimensions are periodic?  i haven't timed this yet but i'm sure it is going to slow down other models which use the cartesian get_close() routine.

Reply:
I remember first seeing this section of code when I started editing location_mod.f90.  Since I believe it was in the original code, I did not run any tests to see what the impact is on accuracy/time if I block the option.   Also I believe CM1 is the only model that uses the file 'assimilation_code/location/threed_cartesian/location_mod.f90'.  mpas is on a cartesian grid but refers to 'xyz_location_mod.f90'.

Reply:
it's possible that this was unintentionally left .true., so for now i'd leave it.  that probably means the "optimized" search code below this case hasn't been fully tested.  we'd like you to make progress on your project so leave this alone and we can test this later.  it's going to take some complicated tests to ensure that the code below (which tries to search only boxes around the base location) actually searches enough of the space to get all locations which are within the given radius.